pkg/driver/driver.go

```diff
@@ -35,120 +31,69 @@ type Driver interface {
-
-	ListSnapshots(_ context.Context) (string, error)
-
-	// ForwardGuestAgent returns if the guest agent sock needs forwarding by host agent.
-	ForwardGuestAgent() bool
-
-	// GuestAgentConn returns the guest agent connection, or nil (if forwarded by ssh).
-	GuestAgentConn(_ context.Context) (net.Conn, error)
-}
-
-type BaseDriver struct {
-	Instance *store.Instance
-
-	SSHLocalPort int
-	VSockPort    int
-	VirtioPort   string
 }
 
-var _ Driver = (*BaseDriver)(nil)
-
-func (d *BaseDriver) Validate() error {
-	return nil
-}
-
-func (d *BaseDriver) Initialize(_ context.Context) error {
-	return nil
-}
-
-func (d *BaseDriver) CreateDisk(_ context.Context) error {
-	return nil
-}
-
-func (d *BaseDriver) Start(_ context.Context) (chan error, error) {
-	return nil, nil
-}
-
-func (d *BaseDriver) CanRunGUI() bool {
-	return false
-}
-
-func (d *BaseDriver) RunGUI() error {
-	return nil
-}
+// GUI defines GUI-related operations.
+type GUI interface {
+	// RunGUI is for starting GUI synchronously by hostagent. This method should be wait and return only after vm terminates
+	// It returns error if there are any failures
+	RunGUI() error
 
-func (d *BaseDriver) Stop(_ context.Context) error {
-	return nil
+	ChangeDisplayPassword(ctx context.Context, password string) error
+	GetDisplayConnection(ctx context.Context) (string, error)
 }
 
-func (d *BaseDriver) Register(_ context.Context) error {
-	return nil
+// SnapshotManager defines operations for managing snapshots.
+type SnapshotManager interface {
+	CreateSnapshot(ctx context.Context, tag string) error
+	ApplySnapshot(ctx context.Context, tag string) error
+	DeleteSnapshot(ctx context.Context, tag string) error
+	ListSnapshots(ctx context.Context) (string, error)
 }
 
-func (d *BaseDriver) Unregister(_ context.Context) error {
-	return nil
+// Registration defines operations for registering and unregistering the driver instance.
+type Registration interface {
+	Register(ctx context.Context) error
+	Unregister(ctx context.Context) error
 }
 
-func (d *BaseDriver) ChangeDisplayPassword(_ context.Context, _ string) error {
-	return nil
-}
+// GuestAgent defines operations for the guest agent.
+type GuestAgent interface {
+	// ForwardGuestAgent returns if the guest agent sock needs forwarding by host agent.
+	ForwardGuestAgent() bool
 
-func (d *BaseDriver) GetDisplayConnection(_ context.Context) (string, error) {
-	return "", nil
+	// GuestAgentConn returns the guest agent connection, or nil (if forwarded by ssh).
+	GuestAgentConn(_ context.Context) (net.Conn, error)
 }
 
-func (d *BaseDriver) CreateSnapshot(_ context.Context, _ string) error {
-	return errors.New("unimplemented")
-}
+// Driver interface is used by hostagent for managing vm.
+type Driver interface {
+	Lifecycle
+	GUI
+	SnapshotManager
+	Registration
+	GuestAgent
 
-func (d *BaseDriver) ApplySnapshot(_ context.Context, _ string) error {
-	return errors.New("unimplemented")
-}
+	GetInfo() Info
```

Comment:
I'm not going to comment on additional setters; they all should drop the `Get` prefix.

```suggestion
	Info() Info
```